python/lsst/daf/butler/cli/cmd/commands.py

```diff
@@ -159,12 +161,61 @@ def query_collections(*args, **kwargs):
 
 
 @click.command()
 @repo_argument(required=True)
 @click.argument('dataset-type-name', nargs=1)
 def remove_dataset_type(*args, **kwargs):
-    """Remove a dataset type definition from a reopsitory."""
-    cli_handle_exception(removeDatasetType, *args, **kwargs)
+    """Remove a dataset type definition from a repository."""
+    cli_handle_exception(script.removeDatasetType, *args, **kwargs)
+
+
+@click.command()
+@repo_argument(required=True)
+@glob_argument(help="GLOB is one or more glob-style expressions that fully or partially identify the "
+                    "dataset types to be queried.")
+@click.option("--collections",
+              help=unwrap("""One or more expressions that fully or partially identify the collections to
+                          search for datasets.If not provided all datasets are returned."""),
+              multiple=True,
+              metavar=typeStrAcceptsMultiple,
+              callback=split_commas)
+@click.option("--where",
+              help=unwrap("""A string expression similar to a SQL WHERE clause. May involve any column of a
+                          dimension table or a dimension name as a shortcut for the primary key column of a
+                          dimension table."""))
+@click.option("--deduplicate",
+              is_flag=True,
+              help=unwrap("""For each result data ID, only yield one DatasetRef of each DatasetType, from the
+                          first collection in which a dataset of that dataset type appears (according to the
+                          order of `collections` passed in).  If used, `collections` must specify at least one
+                          expression and must not contain wildcards."""))
+@click.option("--components",
+              type=click.Choice(["ALL", "NONE", "UNMATCHED"], case_sensitive=False),
+              default="UNMATCHED",
+              show_default=True,
+              metavar="[ALL|NONE|UNMATCHED]",
+              callback=to_upper,
+              help=unwrap("""If UNMATCHED: apply dataset expression patterns to component dataset type names
+                          only if their parent datasets were not matched by the expression. If ALL: apply all
+                          dataset expression patterns to components. If NONE: never apply patterns to
+                          components. Fully-specified component datasets are always included."""))
+@options_file_option()
+def query_datasets(**kwargs):
+    """List the datasets in a repository."""
+    datasets = cli_handle_exception(script.queryDatasets, **kwargs)
+
+    tables = {}
+    for datasetRef in datasets:
+        rows = tables.get(datasetRef.datasetType.name, [])
+        row = OrderedDict(type=datasetRef.datasetType.name, run=datasetRef.run, id=datasetRef.id)
+        row.update(datasetRef.dataId.items())
+        rows.append(row)
+        tables[datasetRef.datasetType.name] = rows
```

Comment:
I don't really like how this bit of code reads the array, updates it, and then assigns it back to the same place. That seems like it should be something that `defaultdict` can handle. The following code results in the same output but there is only one line in the for loop rather than five and it's not continually writing the list back to the dict.

```python
tables = defaultdict(list)
for datasetRef in datasets:
    tables[datasetRef.datasetType.name]].append({"type": datasetRef.datasetType.name, "run": datasetRef.run,
                                      "id": datasetRef.id, **datasetRef.dataId})

for datasetName, rows in tables.items():
    print("")
    print(rows)
    Table(rows, names=rows[0].keys()).pprint_all()
print("")
```
the only annoyance being that `Table` sorts the columns if they aren't an `OrderedDict` and so you need the names here explicitly (there is a comment in astropy that they should fix this when they drop older pythons).

Reply:
@timj there is something about the sorting code that doesn't like sorting (I think) `DimensionElement`s:
```$ butler query-datasets ../pipelines_check/DATA_REPO/ bias

[{'type': 'bias', 'run': 'HSC/calib/gen2/2013-11-03', 'id': 1, Dimension(instrument): 'HSC', Dimension(detector): 10}]
Traceback (most recent call last):
  File "/Users/n8pease/code/lsst/daf_butler/bin/butler", line 28, in <module>
    sys.exit(main())
  File "/Users/n8pease/code/lsst/daf_butler/python/lsst/daf/butler/cli/butler.py", line 311, in main
    return cli()
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/Users/n8pease/code/lsst/daf_butler/python/lsst/daf/butler/cli/cmd/commands.py", line 203, in query_datasets
    Table(rows, names=rows[0].keys()).pprint_all()
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/astropy/table/table.py", line 610, in __init__
    init_func(data, names, dtype, n_cols, copy)
  File "/Users/n8pease/code/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-cb4e2dc/lib/python3.7/site-packages/astropy/table/table.py", line 852, in _init_from_list_of_dicts
    names_from_data = sorted(names_from_data)
  File "/Users/n8pease/code/lsst/daf_butler/python/lsst/daf/butler/core/dimensions/elements.py", line 305, in __gt__
    return self.universe.getElementIndex(self.name) > self.universe.getElementIndex(other.name)
AttributeError: 'str' object has no attribute 'name'
```
I can go debug that if we want(?), or I can use 3 rows in the loop and an `OrderedDict`:
```
    for datasetRef in datasets:
        row = OrderedDict(type=datasetRef.datasetType.name, run=datasetRef.run, id=datasetRef.id)
        row.update(datasetRef.dataId.items())
        tables[datasetRef.datasetType.name].append(row)
```
And then we don't have to do the names annoyance either.

What do you think?

Reply:
That error is from astropy because you have not specified the names as an argument to the Table constructor. Did you remove the names parameter?  Without the names parameter and without an OrderedDict it tries to work out the column names itself but sorts them. With the names parameter it doesn't bother.

You are right that the row is the OrderedDict and tables is the defaultdict so I did miss that last night.

Can't you new example code do the `**datasetRef.dataId` in the OrderedDict constructor can't it? 

Does the following work then?

```python
for datasetRef in datasets:
        row = OrderedDict(type=datasetRef.datasetType.name, run=datasetRef.run, id=datasetRef.id, **datasetRef.dataId)
        tables[datasetRef.datasetType.name].append(row)
```

please add a comment that you are using OrderedDict specifically because current astropy has not yet changed to understand that dicts are ordered. Otherwise people are going to get confused later on.

Reply:
if `rows` is a dict and I pass names to the Table init, it fails as above. See code below.
```
    for datasetRef in datasets:
        row = dict(type=datasetRef.datasetType.name, run=datasetRef.run, id=datasetRef.id)
        row.update(datasetRef.dataId.items())
        tables[datasetRef.datasetType.name].append(row)
```
and this in the command function:
```
    for datasetName, rows in tables.items():
        print("")
        Table(rows, names=rows[0].keys()).pprint_all()
    print("")
```
If I comment out `row.update(datasetRef.dataId.items())` then it does *not* fail.

Reply:
Also I've realized if I make the astropy table in the script function I can do it without the intermediate data structure. Will investigate that briefly.

Reply:
Okay, I think I already said I'm fine with OrderedDict if you use the `**` expansion rather than the separate update.

It looks like they sort the column names even if a names is given and then ignore it to use the names you want.  The problem you are having is that for a dataId the keys of the dataId are not actually strings so you need to convert the dataId to a normal str key dict for sorting to work (or fix the comparison method in the class as you say above -- something we should probably do eventually because sorting keys in a dataId is not completely ridiculous -- not for this ticket).   Not relevant if you stick with OrderedDict along with a note as to why.